File: JS/28-find-the-index-of-the-first-occurrence-in-a-string.js
// string

/**
 * @param {string} haystack
 * @param {string} needle
 * @return {number}
 */
var strStr = function(haystack, needle) {
    function getLPS(pat) {
        let N = Array(pat.length).fill(-1);
        let i = 0, ni = -1;
        while (i < pat.length - 1) {
            if (ni < 0 || pat[i] == pat[ni]) N[++i] = ++ni;
            else ni = N[ni];
        }
        return N;
    }
    let N = getLPS(needle);
    let i = 0, j = 0;
    while (i < haystack.length && j < needle.length) {
        if (j < 0 || haystack[i] == needle[j]) {
            i++; j++;
        } else {
            j = N[j];
        }
    }
    return j == needle.length ? i - j : -1;
};

File: C++/28-find-the-index-of-the-first-occurrence-in-a-string.cpp
#include "header.h"

class Solution {
private:
    vector<int> N;
public:
    void getLPS(string pat) {
        int m = pat.size();
        N = vector<int>(m, 0);
        int ni = -1, i = 0;
        N[0] = ni;
        while (i < m - 1) {
            if (ni < 0 || pat[i] == pat[ni]) {
                i++; ni++;
                N[i] = (pat[i] != pat[ni] ? ni : N[ni]);
            } else {
                ni = N[ni];
            }
        }
    }
    
    int strStr(string haystack, string needle) {
        int n = haystack.size(); int i = 0;
        int m = needle.size(); int j = 0;
        getLPS(needle);
        while (i < n && j < m) {
            if (j < 0 || haystack[i] == needle[j]) {
                i++; j++;
            } else {
                j = N[j];
            }
        }
        return j == m ? i - j : -1; 
    }
};